bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,401 @@
+    private final HttpClient httpClient;
+    private ScheduledFuture<?> pollFuture;
+    private boolean invalidConfiguration = false;
+    Authentication authentication;
+    MessageFactory soapMessageFactory;
+    SOAPFactory soapFactory;
+    private static final int DETECT_POLL_S = 1;
+
+    private final Runnable poller = new Runnable() {
+        @Override
+        public void run() {
+            if (authenticate()) {
+                updateStatus(ThingStatus.ONLINE);
+                poll();
+            }
+        }
+    };
+
+    private static class Authentication {
+        public Authentication(final String privateKey, final String cookie) {
+            this.privateKey = privateKey;
+            this.cookie = cookie;
+        }
+
+        public String privateKey;
+        public String cookie;
+    }
+
+    public DLinkHNAP() {
+        httpClient = new HttpClient();
+    }
+
+    public void start(ScheduledExecutorService scheduler) {
+        try {
+            httpClient.start();
+            pollFuture = scheduler.scheduleWithFixedDelay(poller, 0, DETECT_POLL_S, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.HANDLER_INITIALIZING_ERROR,
+                    "Unexpected internal error.");
+            logger.error("Failed to start http client or scheduler.");
+        }
+    }
+
+    public void stop() {
+        if (pollFuture != null) {
+            pollFuture.cancel(true);
+        }
+        try {
+            httpClient.stop();
+        } catch (Exception e) {
+            logger.error("Failed to stop http client.");
+        }
+    }
+
+    public void reset() {
+        // force re-authentication before the next action.
+        authentication = null;
+
+        // the configuration may be correct.
+        // if it was incorrect, the `authenticate()`-method will set the flag.
+        invalidConfiguration = false;
+    }
+
+    public void handleCommand(final ChannelUID channelUID, final Command command) {
+        if (authenticate()) {
+            handleCommandAuthenticated(channelUID, command);
+        }
+    }
+
+    public void setPin(final String pin) {
+        this.pin = pin;
+    }
+
+    public void setIp(final String ip) {
+        this.url = "http://" + ip + "/HNAP1/";
+    }
+
+    private SOAPElement moduleParameters(final String moduleID) throws SOAPException {
+        return soapFactory.createElement("ModuleID").addTextNode(moduleID);
+    }
+
+    public String queryStringState(final String action, final String name, final String module) {
+        try {
+            final Document document = post(action, newMessage(action, moduleParameters(module)), authentication);
+            return getElementByTagName(document, name);
+        } catch (CommunicationException e) {
+            logger.error("Failed to get quantity (communication error).");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```

Comment:
@Hilbrand is there any issue with performance in calling updateStatus every second?

Reply:
Probably not much. But the status is processed, but probably ignored if not changed. So if a redundant call can be avoided that would save resources. In general in bindings it's checked if the status is changed an only if changed do a status update. But I don't think is consequently done. But if a status update is done every second it might be a good idea to check.

Reply:
Even worse: the behaviour is inconsistent:
When *setting* the state, the code [line 225](https://github.com/openhab/openhab-addons/blob/112de46939ab02101dc51eb0e52e3beec0a26c31/bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java#L225) ignores any `CommunicationException`, but when *querying* the state, the status is set to `Offline` on such an exception.

I suspect that's not good. Should I update the state in
- both cases,
- none of the cases
- or keep it as it is (update only on query)?